### Hipótese 3 (E3)

#### É possível predizer a plataforma de um game baseado em suas vendas?

In [18]:
import numpy as np  
import matplotlib.pyplot as plt  
import pandas as pd

In [19]:
game = pd.read_csv('vgsales.csv', sep=',', encoding='utf-8')

In [20]:
game.head()

Rank                      Name Platform    Year         Genre Publisher  \
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo   
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

In [21]:
X = game.iloc[:, 6:11].values  
y = game.iloc[:, 2].values 

In [22]:
X

array([[4.149e+01, 2.902e+01, 3.770e+00, 8.460e+00, 8.274e+01],
       [2.908e+01, 3.580e+00, 6.810e+00, 7.700e-01, 4.024e+01],
       [1.585e+01, 1.288e+01, 3.790e+00, 3.310e+00, 3.582e+01],
       ...,
       [0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 1.000e-02],
       [0.000e+00, 1.000e-02, 0.000e+00, 0.000e+00, 1.000e-02],
       [1.000e-02, 0.000e+00, 0.000e+00, 0.000e+00, 1.000e-02]])

In [23]:
y

array(['Wii', 'NES', 'Wii', ..., 'PS2', 'DS', 'GBA'], dtype=object)

In [24]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

In [25]:
from sklearn.preprocessing import StandardScaler  
scaler = StandardScaler()  
scaler.fit(X_train)

X_train = scaler.transform(X_train)  
X_test = scaler.transform(X_test)

In [26]:
X_train

array([[-0.18341943, -0.07189133, -0.21874774,  0.00560214, -0.16205656],
       [-0.18341943, -0.22554495, -0.25058312, -0.19612055, -0.24249442],
       [-0.04287661,  0.00493548, -0.21874774, -0.04482853, -0.06924364],
       ...,
       [-0.28882655, -0.28316506, -0.25058312, -0.19612055, -0.31674475],
       [-0.08972421, -0.26395836, -0.25058312, -0.14568988, -0.1929942 ],
       [-0.26540275, -0.22554495, -0.25058312, -0.24655122, -0.29199464]])

In [27]:
X_test

array([[-0.31225035, -0.28316506, -0.12324161, -0.24655122, -0.31055722],
       [-0.13657182, -0.20633825, -0.25058312, -0.19612055, -0.20536925],
       [ 0.5544304 ,  0.21620921, -0.18691236,  0.50990887,  0.3886334 ],
       ...,
       [-0.31225035, -0.28316506, -0.05957085, -0.24655122, -0.29818217],
       [-0.31225035, -0.28316506,  0.89549047, -0.24655122, -0.11255634],
       [-0.21855514, -0.22554495, -0.25058312, -0.19612055, -0.25486947]])

In [28]:
from sklearn.neighbors import KNeighborsClassifier  
classifier = KNeighborsClassifier(n_neighbors=5)  
classifier.fit(X_train, y_train) 

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform')

In [29]:
y_pred = classifier.predict(X_test)

In [30]:
from sklearn.metrics import classification_report, confusion_matrix  
print(confusion_matrix(y_test, y_pred))  
print(classification_report(y_test, y_pred))

[[ 28   0   0   0   0   0   0   0   0   0   0   0   0   1   0   0   0   0
    0   0   0   0   0   0   0   0   1   0]
 [  0  23   0  25   1   0   1   0   0   0   0   3   0   6  13   0   0   2
    1   2   0   4   0   4   1   5   0   0]
 [  0   2   0   5   2   0   0   0   0   0   0   0   0   0   2   0   0   0
    0   0   0   2   0   0   0   0   0   0]
 [  0  45   1 163   1   3   4   0   3   1   0  12   0   4  15   2   3  26
    9   2   0   0   0  81   0  20   7   1]
 [  0   5   1   3   1   2   1   0   0   1   0   0   0   1   2   0   0   0
    0   1   0   2   0   0   0   0   0   0]
 [  0  14   0   1   1 110   4   0   1   1   0   2   0   3   2   0   0   0
    0   1   0   1   0   0   0   0  15   0]
 [  0   3   0   1   3  19  48   0   4   1   0   0   0   1   0   2   0   4
    0   0   0   0   0   0   0   0  27   0]
 [  0   3   0   3   0   0   0   0   0   0   0   0   0   2   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  8   3   0   2   0   6   7   0  17   0   0   0   0   1   0   

/home/joaopaulo/anaconda3/lib/python3.7/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [31]:
success = []

In [ ]:
for i in range(1, 50):  
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(X_train, y_train)
    pred_i = knn.predict(X_test)
    success.append(np.mean(pred_i == y_test))

In [ ]:
success

In [ ]:
plt.figure(figsize=(12, 6))  
plt.plot(range(1, 50), success, color='red', linestyle='dashed', marker='o',  
         markerfacecolor='blue', markersize=10)
plt.title('Taxa acerto de valores K ')  
plt.xlabel('K Valores')  
plt.ylabel('Media de Acerto')

### Conclusão

O melhor resultado que obtemos foi quando se estabelecem algo próximo à 12 vizinhos, tendo uma taxa de acerto média por volta dos 45%. Porém esse valor é baixo e a hipótese acabou sendo refutada, ou seja, não é possível classificar a plataforma de um jogo com precisão baseado apenas nas suas vendas.